src/extract/faster_bottom_up.rs

```diff
@@ -0,0 +1,117 @@
+use super::*;
+
+/// A faster bottom up extractor inspired by the faster-greedy-dag extractor.
+pub struct BottomUpExtractor;
+
+impl Extractor for BottomUpExtractor {
+    fn extract(&self, egraph: &EGraph, _roots: &[ClassId]) -> ExtractionResult {
+        // 1. build map from class to parent nodes
+        let mut parents = IndexMap::<ClassId, Vec<NodeId>>::default();
+        let n2c = |nid: &NodeId| egraph.nid_to_cid(nid);
+
+        for class in egraph.classes().values() {
```

Comment:
nit: can we combine these two loops (e.g. by using the `entry` API)?

Something like:
```rust
for class in egraph.classes().values() {
  for node in &class.nodes {
    for c in &egraph[node].children {
      parents.entry(n2c(c)).or_default().push(node.clone());
    }
  }
}
```

I'd expect it to be roughly the same runtime but worth confirming in the benchmark.

Reply:
For an eclass with no parents, that won't work